[Koh Rui Jie] iP

rjkoh opened the pull request:
### Duke
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you ned to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/rjkoh/ip)
2. double-click it
3. add your tasks
4. let it manage your tasks for you 💩 

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders

-----

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

rjkoh committed:
test 1

rjkoh committed:
Greet, Echo, Exit

rjkoh committed:
Level-2

rjkoh committed:
Level-3: Mark as done

rjkoh committed:
Level 4: ToDos, Events, Deadlines

rjkoh committed:
A-TextUiTesting

rjkoh committed:
Level 5. Handle Errors

rjkoh committed:
Level 6. Delete

rjkoh committed:
style fixes

rjkoh committed:
OOP

rjkoh committed:
test

rjkoh committed:
Level-7

rjkoh committed:
Level-8

rjkoh committed:
level 8

rjkoh committed:
A-MoreOOP

rjkoh committed:
cleanup, enum

rjkoh committed:
organise files into duke package

rjkoh committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

rjkoh committed:
Add JUnit tests

rjkoh committed:
Add JavaDoc Comments

rjkoh committed:
Code comply to coding standard

rjkoh committed:
find tasks by searching keyword

rjkoh committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Code comply to coding standard

rjkoh committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  find tasks by searching keyword

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java

rjkoh committed:
Add javaDocs and follow coding standard

rjkoh committed:
Find tasks by keyword

rjkoh committed:
Improve find and OOP

LuYiting0913 commented on src/main/java/duke/TaskList.java:
```diff
@@ -11,6 +11,7 @@
 public class TaskList {
     private ArrayList<Task> tasks;
```
I like the naming of the variables

LuYiting0913 commented on src/main/java/duke/TaskList.java:
```diff
@@ -11,6 +11,7 @@
     private ArrayList<Task> tasks;
     private int count;
+    private static Ui ui = new Ui();
 
```
I like your JavaDoc. Clear and succinct

LuYiting0913 commented on src/main/java/duke/TaskList.java:
```diff
@@ -60,15 +61,15 @@ public void add(String str, Duke.Type type) {
                                 "Now you have %d tasks in the list.",
                         tasks.get(count - 1).toString(),
                         count));
 
```
Nice comments for different exceptions

rjkoh committed:
Implemented layout for GUI

rjkoh committed:
Respond to user input, with pre-FXML code

rjkoh committed:
Implemented FXML with GUI

rjkoh committed:
Implement CheckStyle and improve style

rjkoh committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Implemented FXML with GUI
  Respond to user input, with pre-FXML code
  Implemented layout for GUI

# Conflicts:
#	src/main/java/duke/TaskList.java

rjkoh committed:
fix style

Nephelite commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,15 @@
+
+public class TaskListTest {
+    @Test
+    public void add_deadline_deadlineAddedWithDate() {
```
Nice job following naming conventions for tests

Nephelite commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Deals with interactions with the user for Duke.
+ */
+public class Ui {
```
I dislike how this class is structured. I suggest having all of Duke's responses listed at the top of the class as constants, and then the class methods would be methods to print those constants. 

Benefits of my suggested change:
This would make editing these greetings much easier to manage as all the strings are at the top of the class.

In addition, if in the future you need to make Ui return the strings it uses in greetings for some reason, this implementation would make such a thing far easier.

Nephelite commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,117 @@
+                    String type = info[0];
+                    String done = info[1];
+                    String name = info[2];
+                    if (type.equals("T")) {
+                        Task todo = new Todo(name);
+                        if (done.equals("1")) {
+                            todo.mark();
+                        }
+                        tasks.add(todo);
+                    } else if (type.equals("D")) {
+                        Task deadline = new Deadline(name, LocalDate.parse(info[3]));
+                        if (done.equals("1")) {
+                            deadline.mark();
+                        }
+                        tasks.add(deadline);
+                    } else if (type.equals("E")) {
+                        String date = info[3];
+                        String time = info[4];
+                        String dt = date + " " + time;
+                        Task event = new Event(name, LocalDateTime.parse(dt,
+                                DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")));
+                        if (done.equals("1")) {
+                            event.mark();
+                        }
+                        tasks.add(event);
+                    } else {
+                        throw new DukeException("Unknown Task type in text file");
+                    }
```
Nice control flow and storage system, so I like the ideas behind it. However, I have a few suggestions to make regarding this structure.

There is far too many nested if statements. I suggest at least abstracting away the ```if(<task>.equals("<task>")``` blocks into a separate method in the task class itself, or maybe even in the ```Parser``` class as a static method. Another layer of abstraction possible is to separate the entire if block here into another method in this class, simply to reduce the levels of nested if statements and to reduce the size of this method.

The system is slightly too optimistic regarding the user. For example, if the user typos and accidentally has // in their task description.

Nephelite commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,71 @@
+
+        // user wants to mark certain task as done
+        } else if (command.startsWith("mark")) {
+            String str = command.replace("mark ", "");
+            int index = Integer.valueOf(str);
+            tasks.mark(index);
+            return "Nice! I've marked this task as done:\n" + tasks.getString(index);
```
I like how the return value is used for feedback here. Quite clever.

Nephelite commented on src/main/java/duke/MissingDescriptionException.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Exception thrown if missing name for tasks
+ */
+public class MissingDescriptionException extends DukeException {
```
I like that you went to create specific new exceptions for your iP.

Nephelite reviewed (changes requested):
I like many of the ideas in this iP. There are some clever bits of code and I like the effort made to create exceptions that extend from DukeException. I also like most of the method names and I think good effort has been made to ensuring coding style conventions have been met. JavaDocs look complete, and there really is not much to complain about.

The only change I would suggest is the one in the ```load()```method of ```Storage``` as it does not utilize Abstraction and overuses nested if statements thus violating coding style contentions.

rjkoh committed:
Add assertions

There are assumptions about the program state.

If they are wrong, there may be a possible bug in the code as the code violates the assumption about how the could should behave.

Let's make use of Java Assert feature to define some assumptions so that they can be verified during runtime.

Using assert will quickly allow us to detect bugs where the assumed code behaviour differs from the actual behaviour, which should not have happened.

rjkoh committed:
Improve code quality

Several methods have poor code quality such as heavy nesting, poor SLAP practice and long implementations.

Such code is difficult to read and maintain especially for others who did not write the code. Low quality code is hard to debug and difficult to understand.

Let's edit the methods, and split methods into the type (addEvent, addTodo, addDeadline) instead of one add method. We also use helper methods when necessary, particularly for lower levels of abstraction.

Doing so is preferable over the current implementation as it makes it more readable, makes the methods shorter and easier to follow.

rjkoh committed:
Merge pull request #1 from rjkoh/branch-A-Assertions

Add assertions

rjkoh committed:
Merge branch 'branch-A-CodeQuality'

* branch-A-CodeQuality:
  Improve code quality

# Conflicts:
#	src/main/java/duke/TaskList.java

rjkoh committed:
Add priority level to tasks

Currently, tasks have no way to keep track of priority.

Users all have differing priority of tasks, some more important than others and it is good to capture this.

Let's implement a priority field in the tasks that will also be written into the text file when saved. This field will be of type enum.

Priority is enum so that users cannot put in unknown priority levels, for safety.

rjkoh committed:
Improve code quality

Errors in javaDocs and others, pointed out by CS2103T bot.

Fixed examples and more.

rjkoh committed:
Add title and bug fixes

Show "Duke" on application header instead of empty and handle IndexOutOfBoundsException in mark and unmark in TaskList.java

rjkoh committed:
Add User Guide

rjkoh committed:
Edit User guide

rjkoh committed:
Update README.md

rjkoh committed:
Edit user guide

rjkoh committed:
Rename UI.png to Ui.png for bot

rjkoh committed:
Rename UI to Ui

rjkoh committed:
Delete old UI.png

rjkoh committed:
Add Ui.png
